src/resource_manager/resource_manager.py

```diff
@@ -0,0 +1,252 @@
+        """
+
+    @abc.abstractmethod
+    def _create_resource(self, resource: AnyResource) -> None:
+        """Abstract method to create a new resource in a given namespace.
+
+        Args:
+            resource: The resource object to create.
+        """
+
+    @abc.abstractmethod
+    def _patch_resource(self, name: str, resource: AnyResource) -> None:
+        """Abstract method to patch an existing resource in a given namespace.
+
+        Args:
+            name: The name of the resource to patch.
+            resource: The modified resource object.
+        """
+
+    @abc.abstractmethod
+    def _list_resource(self) -> typing.List[AnyResource]:
+        """Abstract method to list resources in a given namespace based on a label selector."""
+
+    @abc.abstractmethod
+    def _delete_resource(self, name: str) -> None:
+        """Abstract method to delete a resource from a given namespace.
+
+        Args:
+            name: The name of the resource to delete.
+        """
+
+    def define_resource(
+        self,
+        definition: ResourceDefinition,
+    ) -> AnyResource:
+        """Create or update a resource in kubernetes.
+
+        Args:
+            definition: The ingress definition
+
+        Returns:
+            The name of the created or modified resource.
+
+        Raises:
+            InvalidResourceError: If the generated resource is invalid.
+        """
+        resource_list = self._list_resource()
+        resource = self._gen_resource_from_definition(definition)
+        if not resource.metadata:
+            raise InvalidResourceError("Missing resource metadata.")
+
+        if not resource.metadata.name:
+            raise InvalidResourceError("Missing resource name.")
+
+        resources = [r.metadata.name for r in resource_list if r.metadata]
+        if resource.metadata.name in resources:
+            self._patch_resource(name=resource.metadata.name, resource=resource)
+            logger.info(
+                "%s updated in namespace %s with name %s",
+                self._name,
+                self._namespace,
+                resource.metadata.name,
+            )
+        else:
+            self._create_resource(resource=resource)
+            logger.info(
+                "%s created in namespace %s with name %s",
+                self._name,
+                self._namespace,
+                resource.metadata.name,
+            )
+        return resource
+
+    def cleanup_resources(
+        self,
+        exclude: typing.Optional[AnyResource] = None,
+    ) -> None:
+        """Remove unused resources.
+
+        Args:
+            exclude: The name of resource to be excluded from the cleanup.
+        """
+        for resource in self._list_resource():
+            excluded_resource_name = resource_name(exclude)
+            res_name = resource_name(resource)
+            if res_name is None or excluded_resource_name is None:
```

Comment:
could `res_name` and `excluded_resource_name` be empty strings, i.e. "" ?
This could be `if not res_name or not excluded_resource_name` :)
I'm asking because I'm wondering if `None` represents a special case

Reply:
good point, empty `res_name` should never be the case but it's indeed a possibility we need to cover, I changed all relevant sections to `not res_name`